### PR TITLE
save child pid into state in order to return the correct pid on already_started error

### LINKED ISSUE #102: Horde.Supervisor.start_child returns nil pid when :already_started
When starting a new child with `Horde.Supervisor.start_child/2` and the child is already started, the error `{:error, {:already_started, pid}}` is returned, but pid is always nil.

Having the pid in the returned error is commonly used to perform operation on the process, for example try to start the process -> already started -> GenServer.call(pid, whatever).

### PR DESCRIPTION
**Pull request by xadhoom**
this should fix #102

### CODE REVIEW

File: test/supervisor_test.exs [hunk 1]
```diff
@@ -75,16 +75,19 @@ defmodule SupervisorTest do
...
     end
 
     test "returns error when starting same process twice", context do
       assert {:ok, pid} = Horde.Supervisor.start_child(context.horde_1, context.task_def)
 
-      assert_receive {:process_started, ^pid}
+      assert {:error, {:already_started, child}} =
+               Horde.Supervisor.start_child(context.horde_1, context.task_def)
+
+      assert is_pid(child)
```
Comment (derekkraan) [inline test/supervisor_test.exs]: you can clean up this part by asserting on line 86 as follows:

`assert {:error, {:already_started, ^pid}} = ...`

**Comment by derekkraan** [comment]
Hi @xadhoom, first of all, thanks for the PR. I'm going to think about this one for a couple more days before I come back with comments.

Cheers
Derek

**Comment by xadhoom** [comment]
Mmmh yes, but no one assures me that pid is really a pid :). What if both
returns nil? I'll clean it up but add a type assertion if ok for you.

Il lun 6 mag 2019, 08:56 Derek Kraan <notifications@github.com> ha scritto:

File: test/supervisor_test.exs (hunk 1 above)
Comment (xadhoom) [reply test/supervisor_test.exs]: Mmmh yes, but no one assures me that pid is really a pid :). What if both returns nil? I'll clean it up but add a type assertion if ok for you.

**Comment by xadhoom** [comment]
ok force pushed an update and squashed all commits into one, since is a single feat

**Comment by derekkraan** [comment]
Hi,

First of all, thanks for the PR, and I want to say that the effort is appreciated. I've thought about this, and I think we should support this another way. We should query the relevant Horde instance to ask it about the pid of the process if we need the pid for this error tuple.

Tracking the pids will potentially cause a LOT of churn in the CRDT, to solve what is essentially a race condition, and I want to avoid this.

Will you make the necessary changes to this PR?

Cheers,
Derek

**Comment by xadhoom** [comment]
Don't understand what you mean with "solve what is essentially a race condition"... there're a lot of common cases where you want to start a process in a idempotent way (for example start user/device/whatever tracker when receiving a specific event), and I don't care if started already or not, but I want to call the pid and maybe update it with the new informations.

Also if OTP provides the pid to such APIs, there's a lot of sense in providing it :) (I assume who designed OTP is much smarter and have seen many more use cases than me)

Probably the same can be accomplished using a registry, but is more convoluted (ask registry, start if not there, what if 2 events are received at almost same time? you'll need to handle already_started error anyway)

About CRDT churn... why? It may be a problem if your process dies badly and get restarted often, which should not be the normal running condition. In any other case the data is put in the CRDT at creation, so not creating additional CRDT work (same for deletion). Or I'm missing something?

Said that, the PR can be updated, but I need some time to figure out how to find the owning Horde instance and add proper APIs (this will be indeed slower than keeping a copy of the information on local state like all other data without having to to a remote call to another node)

What do you think?

**Comment by derekkraan** [comment]
Hi,

Just a quick note, I'm still thinking about this PR, just not sure yet what to do. I want to ge it right. Just so you don't think I'm ignoring this PR. :stuck_out_tongue: 

Derek

**Comment by derekkraan** [comment]
Specifically I'm thinking about our conversation here: https://github.com/derekkraan/horde/issues/54 and whether it's maybe a better idea to ignore `id` (we'd still need an internal `id` of course, but we could generate that). This would bring us into better alignment with `DynamicSupervisor` and also solve the issue of this PR.

**Comment by derekkraan** [comment]
If we move to ignoring `id`, like DynamicSupervisor does, then we have to track the `pid` anyways, otherwise we can't get things like `terminate_child` to work :man_facepalming:

**Comment by derekkraan** [comment]
I'm going to merge this into another branch and do some experiments on it there, see if we can become more like DynamicSupervisor on the whole.

**Pull request merged**